### PR TITLE
Improve errors in irmin-pack/unix

### PR DESCRIPTION
**Pull request by Ngoguey42**

### CODE REVIEW

**Comment by Ngoguey42** [comment]
I've rebase. New tests fail:
```
> [FAIL]        dict-files                    0   dict.
  [FAIL]        dict-files                    1   RO dict.
  [FAIL]        pack-files                    3   close readonly.
  [FAIL]        pack-files                    4   readonly sync, index flush.
  [FAIL]        branch-files                  0   branch.
  [FAIL]        branch-files                  1   branch close.
  [FAIL]        existing stores               0   Test index reconstruction.
  [FAIL]        existing stores               1   Test integrity check.
  [FAIL]        existing stores               2   Test integrity check for inodes.
  [FAIL]        version-bump                  0   test_RO_no_version_bump.
  [FAIL]        version-bump                  1   test_RW_no_version_bump.
  [FAIL]        version-bump                  2   test_RW_version_bump.

  ```

**Comment by Ngoguey42** [comment]
Let's not block this PR because of the failing tests. Let's deal with these after

File: src/irmin-pack/unix/errors_base.ml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  | `Invalid_layout
+  | `Corrupted_legacy_file
+  | `Pending_flush
+  | `Rw_not_allowed ]
+[@@deriving irmin ~pp]
+
+(* This exception does not contain [`Ro_not_allowed].
+
+   We can't use polyval inclusion because repr doesn't support it *)
+type error' =
```
Comment (icristescu) [inline src/irmin-pack/unix/errors_base.ml:42]: I don't understand why having `error'` (the patch compiles https://github.com/icristescu/irmin/commit/989ad1cd7b899bbbed596742ddbe3efe8ce951b0 but I maybe missed something). 

Also, why is Io_mis not included?

File: test/irmin-pack/test_inode.ml [hunk 2]
```diff
@@ -44,17 +44,19 @@ struct
     Irmin_pack.Inode.Make_internal (Conf) (Schema.Hash) (Key) (Node)
 
   module Io = Irmin_pack_unix.Io.Unix
+  module Errs = Irmin_pack_unix.Errors.Make (Io)
```
Comment (icristescu) [inline test/irmin-pack/test_inode.ml:47]:
```suggestion
```
this is also exposed by common

**Comment by icristescu** [comment]
Can you cherry-pick https://github.com/mirage/irmin/commit/8729e0119b4c5b7b9ed14901d03f167245c0b320 ? it makes the tests pass (the issue was due to a poor isolation between them)

File: src/irmin-pack/unix/errors_base.ml (hunk 1 above)
Comment (icristescu) [reply src/irmin-pack/unix/errors_base.ml:42]: Ok, thanks!

**Review by icristescu — APPROVED** [review]

File: src/irmin-pack/unix/io_intf.ml [hunk 3]
```diff
@@ -127,23 +109,23 @@ module type S = sig
...
   val write_exn : t -> off:int63 -> string -> unit
   (** [write_exn t ~off b] writes [b] to [t] at offset [off].
 
-      Raises [Io_error]
+      Raises [Errors.Io_error]
 
       Also raises backend-specific exceptions (e.g. [Unix.Unix_error] for the
       unix backend). *)
+
+  val raise_misc_error : misc_error -> 'a
```
Comment (tomjridge) [inline src/irmin-pack/unix/io_intf.ml:125]: Maybe rename as "raise_misc_error_as_unix_error"? A bit unwieldy but clearer.

**Pull request merged**